### PR TITLE
sync: reduce memory allocation when write into files

### PR DESCRIPTION
**Pull request by 201341**
Signed-off-by: swj <1186093704@qq.com>

### CODE REVIEW

**Comment by SandyXSD** [comment]
@201341 Could you provide some detailed explanation?

**Comment by 201341** [comment]
> @201341 Could you provide some detailed explanation?

For file2file, io.CopyBuffer -> readfrom->io.Copy, it doesn't use the bufpool.

**Comment by SandyXSD** [comment]
*os.File has the method `ReadFrom`, why do you want to convert its type to WriteCloser?

**Comment by 201341** [comment]
> *os.File has the method `ReadFrom`, why do you want to convert its type to WriteCloser?

In order to use the buffer, not to alloc the memory, we need to convert to WriteCloser.

**Comment by davies** [comment]
@201341 io.CopyBuffer() will use copy_file_range (zero copy or in kernel) in linux, and fallback to io.Copy() in other OS, where do you saw the allocation?

**Comment by 201341** [comment]
> @201341 io.CopyBuffer() will use copy_file_range (zero copy or in kernel) in linux, and fallback to io.Copy() in other OS, where do you saw the allocation?

From pprof, `io.CopyBuffer->io.Copy()`  takes up most of the memory, And I see the code `io.Copy`  allocate the new memory!

**Comment by davies** [comment]
What's the OS and version you are using?

**Comment by 201341** [comment]
> What's the OS and version you are using?

The version is Centos7(kernel 3.x), And it seems not support the `copy_file_range`.

**Comment by SandyXSD** [comment]
This change makes sense since `ReadFrom` of File requires the `src` to be a *File as well to use `copy_file_range`, while in JuiceFS usually the src is bytes.Reader (except for sync/backup).

```
src, ok := r.(*File)
if !ok {
    return 0, false, nil
}
```

**Comment by davies** [comment]
I see, the pool is not used when the `in` is not a `File`, or copy_file_range is not available.

We should check the difference when src and dst are disk and copy_file_range is available.

**Comment by SandyXSD** [comment]
@201341 It looks like this commit won't change the behavior at all, cuz `io.WriteCloser(f)` still contains the `ReadFrom` method. Could you double check to confirm if it fixes the issue of `io.Copy`?

**Comment by 201341** [comment]
@SandyXSD Yes, it doesn't change the behavior. I have changed the code. Now it can use the buffer, But I don't know how to check the copy_file_range

**Comment by SandyXSD** [comment]
Cool, and we need a method to check kernel version. After a quick look at the issue you linked, it looks that the [bug](https://bugzilla.redhat.com/show_bug.cgi?id=1803162) is only NFS related?

File: pkg/object/object_storage.go [hunk 1]
```diff
@@ -42,6 +43,10 @@ type File interface {
 	Mode() os.FileMode
 }
 
+type OnlyWriter struct {
```
Comment (davies) [inline pkg/object/object_storage.go:46]: lower case

**Review by SandyXSD — APPROVED** [review]

File: pkg/object/object_storage.go (hunk 1 above)
Comment (201341) [reply pkg/object/object_storage.go]: fix it.

**Review by davies — APPROVED** [review]

**Comment by davies** [comment]
@201341 Thanks! This patch disable the copy_file_range when both source and destination are same file system, we will find a way to enable it when it's safe.

**Comment by penglaixy99** [comment]
From golang official webpage:   https://pkg.go.dev/io#CopyBuffer
CopyBuffer is identical to Copy except that it stages through the provided buffer (if one is required) rather than allocating a temporary one. If buf is nil, one is allocated; otherwise if it has zero length, CopyBuffer panics.

If either src implements WriterTo or dst implements ReaderFrom, buf will not be used to perform the copy.

**Pull request merged**